p4-16/spec/P4-16-spec.mdk

```diff
@@ -1190,8 +1190,11 @@ P4 offers a very simple module system.  Modules enable programmers to
 decompose a program into multiple source files and libraries.
 
 In the compilation process the preprocessing step is executed first.
-After preprocessing the modules are loaded.  Modules offer a
-functionality similar to `#include` directives:
+After preprocessing the modules are loaded.  Before a module is loaded
+a *new* preprocessor is applied to its source file (no preprocessor
+definitions are inherited from previous preprocessor invocations).  If
+a module is imported multiple times it is preprocessed anew each time.
+Modules offer a functionality similar to `#include` directives:
```

Comment:
what if module A imports B and module B imports A?  Is that allowed?

what if mulitple modules A, B, and C all import X?  Is it permissible to parse X once and import it into all of them, or do we have to reparse each time (getting different values for macros like __TIME__`)

Reply:
What if a module instantiates some extern, and the module is then imported into multiple other modules?  Does that result in a separate instance for every import, or just one instance that is shared?

Reply:
Technically the spec does not promise any CPP macros except the ones defined by the user without any arguments, so `__TIME__` is not guaranteed to work even with just the preprocessor. It would be nice to get rid of the preprocessor - then we could answer these questions much easier - there are no macros in modules, only in the preprocessor.

But I kept the preprocessor for three reasons:
- backwards compatibility
- more expressivity in the preprocessor, some of which may be useful
- we don't have a replacement for `#ifdef/#else`

The module system is not really designed for performance, so I don't personally mind parsing X three times. We should produce the most natural result the users would expect. For circular inclusions this could probably be an error message.

This could be a problem indeed if two libraries have private modules named "a", which they both import. With the import scheme proposed in this draft the second library would break the first one. This would suggest that import pathnames to be searched starting with the "current importing directory" rather than the global import path.

Reply:
Yes, the extern is a trickier question. This only is important for top-level externs really.
The extern will not be visible from the outside, so none of its methods can ever be called by user code. I am tempted to say that in this case the behavior is platform-dependent. I don't know if there are good uses for this capability.